Add group membership discovery

maximilianriemensberger opened the pull request:
Issue #27 

(based on top of PR #32)

Add GD and GP pck commands and the corresponding status messages.  Also add convenience methods to ModuleConnection.

alengwenus commented on pypck/module.py:
```diff
@@ -305,6 +305,120 @@ def oem_text(self) -> List[str]:
+        """
+        if isinstance(inp, inputs.ModStatusGroups):
+            if inp.dynamic:
+                self._dynamic_groups = inp.groups
```
The `self.dynamic_groups_known_event.set()` must be invoked here.
Otherwise we repeat requesting group membership again and again.

alengwenus commented on pypck/module.py:
```diff
@@ -305,6 +305,120 @@ def oem_text(self) -> List[str]:
+            if inp.dynamic:
+                self._dynamic_groups = inp.groups
+            else:
+                self._static_groups = inp.groups
```
See above but with `self.static_groups_known.set()`.

alengwenus commented on pypck/module.py:
```diff
@@ -305,6 +305,120 @@ def oem_text(self) -> List[str]:
+        else:
+            self.static_groups_known.set()
+
+    def timeout_dynamic_groups(self, failed: bool = False) -> None:
```
This has to be a coroutine.

alengwenus commented on pypck/module.py:
```diff
@@ -305,6 +305,120 @@ def oem_text(self) -> List[str]:
+    def timeout_dynamic_groups(self, failed: bool = False) -> None:
+        """Is called on dynamic group membership request timeout."""
+        if not failed:
+            self.addr_conn.send_command(
```
`send_command()` is a coroutine and must be awaited here.

alengwenus commented on pypck/module.py:
```diff
@@ -305,6 +305,120 @@ def oem_text(self) -> List[str]:
+            else:
+                self._static_groups = inp.groups
+
+    def timeout_static_groups(self, failed: bool = False) -> None:
```
This has to be a coroutine.

alengwenus commented on pypck/module.py:
```diff
@@ -305,6 +305,120 @@ def oem_text(self) -> List[str]:
+    def timeout_static_groups(self, failed: bool = False) -> None:
+        """Is called on static group membership request timeout."""
+        if not failed:
+            self.addr_conn.send_command(
```
`send_command()` is a coroutine and must be awaited here.

alengwenus commented on pypck/module.py:
```diff
@@ -305,6 +305,120 @@ def oem_text(self) -> List[str]:
+                False, PckGenerator.request_group_membership_dynamic()
+            )
+        else:
+            self.static_groups_known.set()
```
Typo. Should be: `self.dynamic_groups_known.set()`

alengwenus commented on pypck/module.py:
```diff
@@ -305,6 +305,120 @@ def oem_text(self) -> List[str]:
+    async def request_dynamic_groups(self) -> List[LcnAddr]:
+        """Request dynamic group membership from a module."""
+        await self.addr_conn.conn.segment_scan_completed_event.wait()
+        self.static_groups_known.clear()
+        self.static_groups_trh.activate()
+        await self.static_groups_known.wait()
+        return self.static_groups
```
Copy paste? :smirk:  Should all be `dynamic_`...

alengwenus commented on pypck/module.py:
```diff
@@ -305,6 +305,120 @@ def oem_text(self) -> List[str]:
+
+    async def cancel_dynamic_groups(self) -> None:
+        """Cancel static groups request task."""
+        await self.static_groups_trh.cancel()
```
`await self.dynamic_groups_trh.cancel()`

alengwenus commented on pypck/module.py:
```diff
@@ -1436,6 +1556,21 @@ def oem_text(self) -> List[str]:
+
+    @property
+    def groups(self) -> List[LcnAddr]:
+        """Return dynamic group membership."""
```
Should be static and dynamic.

alengwenus commented on pypck/module.py:
```diff
@@ -305,6 +305,120 @@ def oem_text(self) -> List[str]:
+        await self.static_groups_trh.cancel()
+
+    async def cancel(self) -> None:
+        """Cancel all name, comment and OEM text request tasks."""
```
Wrong docstring,

alengwenus commented on pypck/module.py:
```diff
@@ -305,6 +305,120 @@ def oem_text(self) -> List[str]:
+
+    @property
+    def static_groups(self) -> List[LcnAddr]:
+        """Return stored name."""
```
Wrong docstring.

alengwenus commented on pypck/module.py:
```diff
@@ -305,6 +305,120 @@ def oem_text(self) -> List[str]:
+
+    @property
+    def dynamic_groups(self) -> List[LcnAddr]:
+        """Return stored name."""
```
Wrong docstring.

alengwenus commented on pypck/module.py:
```diff
@@ -305,6 +305,120 @@ def oem_text(self) -> List[str]:
+
+    @property
+    def groups(self) -> List[LcnAddr]:
+        """Return stored name."""
```
Wrong docstring.

alengwenus reviewed (changes requested):
Great! I'm impressed with how deep you are into the code. 👍 
If you implement the requested changes, I think it should work.

maximilianriemensberger commented:
@alengwenus Thanks for the detailed review. I'll address your comments. And it looks like some tests wouldn't do any harm either :wink:.

maximilianriemensberger commented:
I addressed your comments, but I'll have to familiarize myself with the integration test infrastructure for adding the tests. So you might want to wait until I got those done.

maximilianriemensberger commented:
Did some tests. Ready to go from my POV.

alengwenus commented on tests/test_connection.py:
```diff
@@ -1,8 +1,9 @@
 import asynctest
 import pytest
-from pypck.connection import PchkAuthenticationError, PchkLicenseError
+from pypck.connection import cancel, PchkAuthenticationError, PchkLicenseError
```
`cancel` is not used
Adding the tests folder to the linters should be one of the next things we should do. 😏

maximilianriemensberger replied:
Sounds like a good idea.

alengwenus commented on pypck/pck_commands.py:
```diff
@@ -74,6 +74,14 @@ class PckParser:
     )
 
+    # Pattern to parse the static and dynamic group membership status
+    PATTERN_STATUS_GROUPS = re.compile(
+        r"=M(?P<seg_id>\d{3})(?P<mod_id>\d{3})\.G(?P<kind>[DP])"
+        r"(?:(?P<g1>\d{3}))?(?:(?P<g2>\d{3}))?(?:(?P<g3>\d{3}))?(?:(?P<g4>\d{3}))?"
+        r"(?:(?P<g5>\d{3}))?(?:(?P<g6>\d{3}))?(?:(?P<g7>\d{3}))?(?:(?P<g8>\d{3}))?"
+        r"(?:(?P<g9>\d{3}))?(?:(?P<g10>\d{3}))?(?:(?P<g11>\d{3}))?(?:(?P<g12>\d{3}))?"
+    )
```
Unfortunately, I realized that this pattern is **not** correct.
After GP/GD there must be the number of groups as a three-digit value. As far as I can see this is always 12 for GP and 8 for GD (no matter how many groups are actually programmed).
However, you should add `(?P<groups>\d{3})` at the end of first line in the expression and update everything else accordingly.
Sorry for that!

maximilianriemensberger replied:
Ah ... I was already thinking something is special about groups 8, 10, 12 or something is weird about my modules. Most of my UP modules have 8 most of my din rail module have 10 as the number of groups.  The number of static groups is indeed 12 for all of them.

alengwenus commented on pypck/module.py:
```diff
@@ -305,6 +305,122 @@ def oem_text(self) -> List[str]:
+        if isinstance(inp, inputs.ModStatusGroups):
+            if inp.dynamic:
+                self._dynamic_groups = inp.groups
+                self.dynamic_groups_known.set()
```
I rechecked it. To avoid an assertion error we have to check if the event is already set. ALso we have to cancel the corresponding TimeoutRetryHandler:
```python3
if not self.dynamic_groups_known.is_set():
    self.dynamic_groups_known.set()
await self.cancel_dynamic_groups()
```

maximilianriemensberger replied:
Hm `self.dynamic_groups_known.set()` should do nothing id the Event is already set.  At least the docs page does not indicate that it is not allowed to call `set()` if `is_set()` is already.

The cancel definitely should be called. So I would assume the following should work:

```
self.dynamic_groups_known.set()
await self.cancel_dynamic_groups()
```

I tried to repro the assert but couldn't trigger it. Otherwise I'd add a test for it.

alengwenus replied:
Yes, you are right. Prior to Events I used Futures. Future.set_result() raises an exception if already set. Events don't.

alengwenus commented on pypck/module.py:
```diff
@@ -305,6 +305,122 @@ def oem_text(self) -> List[str]:
+                self.dynamic_groups_known.set()
+            else:
+                self._static_groups = inp.groups
+                self.static_groups_known.set()
```
Same as above, but static here:
```python3
if not self.static_groups_known.is_set():
    self.static_groups_known.set()
await self.cancel_static_groups()
```

alengwenus commented:
According to [https://www.lcn.eu/lcn-tech/pchk-historie/](https://www.lcn.eu/lcn-tech/pchk-historie/) the GP/GD command is only available for PCHK version >=3.00 and modules with firmware >=19090D. The latter we could check to speed up execution for older modules (maybe raise an exception to notify the user?).
I opened an issue for this (#40) . We shouldn't address the firware check here to keep the complexity of this PR low.

alengwenus reviewed (approved):
Looks good! 👍
